sdk/webpubsub-socketio-extension/src/SIO/components/negotiate.ts

```diff
@@ -0,0 +1,122 @@
+      const tokenResponse = await serviceClient.getClientAccessToken(negotiateOptions);
+      const url = new URL(tokenResponse.baseUrl);
+      const message: NegotiateResponse = {
+        endpoint: url.origin,
+        path: url.pathname,
+        token: tokenResponse.token,
+      };
+      writeResponse(res, 200, message);
+      debug("negotiate, finished");
+    } catch (e) {
+      writeResponse(res, 500, { message: "Internal Server Error" });
+      debug(`negotiate, error: ${e.message}`);
+    }
+  };
+}
+
+/**
+ * Get a `ConfigureNegotiateOptions` used by `negotaite` to enable passport authentication.
+ * Using this middleware, the user id of passport will be put into the response for negotiate request as a part of JWT token.
+ * When the Socket.IO client connects to the server, the user id in JWT token will be extracted and the corresponding passport will be restored and `socket.request.user` will be available.
+ *
+ * @param store - the storage for `express-session`
+ * @param assignProperty - the property name of passport object in `socket.request`. Default value is `user`. Reference: https://www.jsdocs.io/package/@types/passport#AuthenticateOptions.assignProperty
+ * @returns
+ */
+export function usePassport(store: Store, assignProperty = "user"): ConfigureNegotiateOptions {
+  if (!assignProperty || assignProperty.length === 0) {
+    throw new Error("Valid assignProperty is required.");
+  }
+
+  return async (req: ExpressRequest) => {
+    const query = parse(req.url || "", true).query;
+    const expirationMinutes = Number(query["expirationMinutes"] ?? 600);
+
+    const sessionId = req.headers.cookie.slice("connect.sid=s%3A".length).split(".")[0];
+
+    const session = JSON.parse(store["sessions"][sessionId]);
+    if (session["passport"]) {
+      const passportUserId = session["passport"][assignProperty];
+      return { userId: passportUserId, expirationTimeInMinutes: expirationMinutes };
+    }
+    return { expirationTimeInMinutes: expirationMinutes };
+  };
+}
+
+/**
+ * Get a Express which cooperates with `usePassport` to restore passport in `socket.request`.
+ *
+ * @param assignProperty - the property name of passport object in `socket.request`. Default value is `user`. Reference: https://www.jsdocs.io/package/@types/passport#AuthenticateOptions.assignProperty
+ * @returns
+ */
+export function restorePassport(assignProperty = "user") {
+  return (request: ExpressRequest, response: ExpressResponse, next: any) => {
+    try {
+      const passportUserId = JSON.parse(request["claims"].userId);
+      request["session"] = { passport: {} } as unknown as Session;
+      request["session"]["passport"][assignProperty] = passportUserId;
+    } catch (e) {
+      debug(`restorePassport, error: ${e.message}`);
+    }
+    next();
+  };
+}
+
+/**
+ * Get a Express which cooperates with `negotiate` to restore claims in `socket.request`.
+ *
+ * @returns
+ */
+export function restoreClaims() {
```

Comment:
Do we need to restore `userId` to `req.userId`?

Reply:
discussed offline. better not